pyvisa/resources/async.py

```diff
@@ -0,0 +1,183 @@
+
+
+@method_of('MessageBasedResource')
+@coroutine
+def _async_read_chunk(self, size=None):
+    with self.ignore_warning(constants.VI_SUCCESS_SYNC):
+        self.start_async_read(size)
+
+    while True:
+        yield
+        msg, job_status = self._check_async_read(timeout=0)
+        if msg is not None:
+            return msg, job_status
+
+
+@method_of('MessageBasedResource')
+@coroutine
+def _async_read_raw(self, size=None):
+    ret = bytearray()
+    job_status = loop_status = constants.StatusCode.success_max_count_read
+    with self.ignore_warning(constants.VI_SUCCESS_DEV_NPRESENT,
+                             constants.VI_SUCCESS_MAX_CNT):
+        while job_status == loop_status:
+            logger.debug('%s - reading %d bytes (last status %r)',
+                         self._resource_name, size, job_status)
+            msg, job_status = yield from self._async_read_chunk(size)
+            ret.extend(msg)
+    return bytes(ret)
+
+
+@method_of('MessageBasedResource')
+@coroutine
+def async_read_raw(self, size=None):
+    """Read the unmodified string asynchronously
+
+    In contrast to async_read(), no termination characters are stripped.
+
+    Use ``yield from`` to get the value from within another coroutine or
+    generator::
+        msg = yield from resource.async_read_raw()
+
+    :param size: The chunk size to use when reading the data.
+
+    :rtype: bytes
+
+    Only available for Python 3.3 and above.
+    """
+    return bytes((yield from self._async_read_raw(size)))
+
+
+@method_of('MessageBasedResource')
+@coroutine
+def async_read(self, termination=None, encoding=None):
+    """Read a string from the device asynchronously.
+
+    Use ``yield from`` to get the value from within another coroutine or
+    generator::
+        msg = yield from resource.async_read()
+
+    Reading stops when the device stops sending (e.g. by setting
+    appropriate bus lines), or the termination characters sequence was
+    detected.  Attention: Only the last character of the termination
+    characters is really used to stop reading, however, the whole sequence
+    is compared to the ending of the read string message.  If they don't
+    match, a warning is issued.
+
+    All line-ending characters are stripped from the end of the string.
+
+    :rtype: str
+
+    Only available for Python 3.3 and above.
+    """
+    enco = self._encoding if encoding is None else encoding
+
+    if termination is None:
+        termination = self._read_termination
+        message = (yield from self._async_read_raw()).decode(enco)
+    else:
+        with self.read_termination_context(termination):
+            message = (yield from self._async_read_raw()).decode(enco)
+
+    if not termination:
+        return message
+
+    if not message.endswith(termination):
+        warnings.warn("read string doesn't end with "
+                      "termination characters", stacklevel=2)
+
+    return message[:-len(termination)]
```

Comment:
I would prefer to be complete here and hence have also `async_read_bytes`, `async_read_ascii_values`, `async_read_binary_values`. I am not sure what is the best way to limit code duplication here however.